tenants/tenants.go

```diff
@@ -0,0 +1,858 @@
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("cannot read CA certificate file for tenant %q: %v", t.Name, err))
+
+				return err
+			}
+		}
+
+		var (
+			block *pem.Block
+			rest  []byte = t.MTLS.RawCA
+			cert  *x509.Certificate
+		)
+
+		for {
+			block, rest = pem.Decode(rest)
+
+			if block == nil {
+				skip.Log("tenant", t.Name, "err", "failed to parse CA certificate PEM")
+
+				break
+			}
+
+			cert, err = x509.ParseCertificate(block.Bytes)
+
+			if err != nil {
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("failed to parse CA certificate: %v", err))
+
+				break
+			}
+
+			t.MTLS.cas = append(t.MTLS.cas, cert)
+
+			if len(rest) == 0 {
+				break
+			}
+		}
+	}
+
+	return nil
+}
+
+// Load OIDC auth information for a tenant.
+func loadOIDCConf(tCfg *tenantsConfig, t *tenant) error {
+	skip := level.Warn(log.With(tCfg.logger, "msg", "skipping invalid tenant"))
+
+	var err error
+
+	if t.OIDC != nil {
+		if t.OIDC.IssuerCAPath != "" {
+			t.OIDC.IssuerRawCA, err = ioutil.ReadFile(t.OIDC.IssuerCAPath)
+			if err != nil {
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("cannot read issuer CA certificate file for tenant : %q %v", t.Name, err))
+
+				return err
+			}
+		}
+
+		if len(t.OIDC.IssuerRawCA) != 0 {
+			block, _ := pem.Decode(t.OIDC.IssuerRawCA)
+
+			if block == nil {
+				skip.Log("tenant", t.Name, "err", "failed to parse issuer CA certificate PEM")
+
+				return err
+			}
+
+			cert, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("failed to parse issuer certificate: %v", err))
+
+				return err
+			}
+
+			t.OIDC.issuerCA = cert
+		}
+	}
+
+	return nil
+}
+
+// Onboard a tenant.
+func newTenant(cfg *config, tCfg *tenantsConfig, r *chi.Mux, t *tenant) {
+	retryCount := 0
+
+	retryCounterMetric, _ := registerTenantRetryMetric(tCfg)
+
+	for {
+		// Calculate wait time before rerying a failed tenant registration.
+		waitTime := expBackOffWaitTime(retryCount)
+		time.Sleep(time.Duration(waitTime) * time.Millisecond)
+
+		rateLimits := tenantRateLimits(tCfg, t)
+
+		oidcConf, mTLSConf, err := tennatAuthN(t)
+		if err != nil {
+			level.Error(tCfg.logger).Log("msg", "tenant must specify either an OIDC or an mTLS configuration", t.Name, err)
+			retryCounterMetric.With(prometheus.Labels{"tenant": t.Name}).Inc()
+			retryCount++
+
+			continue
+		}
+
+		var (
+			authN       authentication.Middleware
+			authZ       rbac.Authorizer
+			oidcHandler http.Handler
+		)
+
+		if len(oidcConf.Tenant) > 0 {
+			// Create OIDC middleware for a tenant.
+			oidcHandler, authN, err = authentication.NewOIDC(tCfg.logger, "/oidc/"+t.Name, *oidcConf)
+			if err != nil {
+				level.Error(tCfg.logger).Log("msg", "tenant failed to register. retrying ..", t.Name, err)
+				retryCounterMetric.With(prometheus.Labels{"tenant": t.Name}).Inc()
+				retryCount++
+
+				continue
+			}
+		}
+
+		if oidcHandler != nil {
+			r.Mount("/oidc/"+t.Name, oidcHandler)
+		}
```

Comment:
Could this move to the end of preceding `ìf` block? I guess it would make it more apparent that it belongs to OIDC handler initiation part.